server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -680,72 +689,156 @@ public String toString() {
                 '}';
     }
 
-    void sendFiles(Store store, StoreFileMetaData[] files, Supplier<Integer> translogOps) throws Exception {
-        ArrayUtil.timSort(files, Comparator.comparingLong(StoreFileMetaData::length)); // send smallest first
-        final LocalCheckpointTracker requestSeqIdTracker = new LocalCheckpointTracker(NO_OPS_PERFORMED, NO_OPS_PERFORMED);
-        final AtomicReference<Tuple<StoreFileMetaData, Exception>> error = new AtomicReference<>();
-        final byte[] buffer = new byte[chunkSizeInBytes];
-        for (final StoreFileMetaData md : files) {
-            if (error.get() != null) {
-                break;
+    void sendFiles(Store store, StoreFileMetaData[] files, Supplier<Integer> translogOps, ActionListener<Void> listener) {
+        final MultiFileSender multiFileSender = new MultiFileSender(store, files, translogOps);
+        resources.add(multiFileSender); // need to register to the resource list so we can clean up if the recovery gets cancelled.
+        final ActionListener<Void> wrappedListener = ActionListener.wrap(
+            r -> {
+                if (Assertions.ENABLED) {
```

Comment:
can we do this differently? can you move it out to a method and call `assert method(args);`